Prototype print flags get

juanvallejo opened the pull request:
@deads2k @soltysh
Opening this prototype in private in order to address a few issues we might face while implementing PrintFlags for the humanreadable printer at the command level.

juanvallejo commented on pkg/kubectl/cmd/resource/get.go:
```diff
@@ -196,12 +214,20 @@ func (options *GetOptions) Complete(f cmdutil.Factory, cmd *cobra.Command, args
 	}
 
+	options.HumanPrintFlags.WithNamespace = options.AllNamespaces
+	options.HumanPrintFlags.AbsoluteTimestamps = options.Watch
```
unsure about setting these here.

deads2k replied:
> unsure about setting these here.

ugly, but acceptable.

juanvallejo commented on pkg/kubectl/cmd/resource/get.go:
```diff
@@ -320,145 +345,51 @@ func (options *GetOptions) Run(f cmdutil.Factory, cmd *cobra.Command, args []str
 
-	sorting, err := cmd.Flags().GetString("sort-by")
+	showKind := options.ShowKind || resource.MultipleTypesRequested(args) || cmdutil.MustPrintWithKinds(objs, infos, nil)
+	options.HumanPrintFlags.ShowLabels = &showKind
```
ShowKind gets set through a flag, but we end up overriding it here anyway, since it appears that we want to show it regardless of what the flag says when there is more than one resource kind

deads2k replied:
This looks wrong.  Setting showlabels based the kind value?

juanvallejo replied:
Thanks, meant to set `options.HumanPrintFlags.ShowKind = &showKind`. Updated

juanvallejo commented on pkg/kubectl/cmd/resource/get.go:
```diff
@@ -320,145 +345,51 @@ func (options *GetOptions) Run(f cmdutil.Factory, cmd *cobra.Command, args []str
-		}
+	// TODO(juanvallejo): handling this inside of humanprinter_flags.go causes an import cycle
+	// figure out how to attach these handlers without resorting to each command having to do it.
+	printersinternal.AddHandlers(printer.(*printers.HumanReadablePrinter))
```
So far everything works as expected, however I'm stuck having to add print handlers to the printer here, since adding them in `humanreadable_flags.go` causes an import cycle.

Solution would be to move the handlers out of `pkg/printers/internalversion` and into `pkg/printers`

deads2k replied:
> Solution would be to move the handlers out of pkg/printers/internalversion and into pkg/printers

Wrong solution.  Make a new package.

juanvallejo commented on pkg/kubectl/cmd/resource/get.go:
```diff
@@ -320,145 +345,51 @@ func (options *GetOptions) Run(f cmdutil.Factory, cmd *cobra.Command, args []str
-			} else {
-				resourceName = "none"
-			}
+	list := api.List{
```
I am taking all of the infos I receive, and aggregating them in an api list object.
I then decode that object using the universal decoder in order to obtain a runtime.Object, which the `PrintObj` method receives.

deads2k replied:
> I am taking all of the infos I receive, and aggregating them in an api list object.

This sounds wrong.  We were visiting before, weren't we?

juanvallejo replied:
We were looping over a slice of (sorted) runtime.Objects before.
Then we were obtaining a new PrinterForMapping for each object, and passing it to that printer's PrintObj method.

We could still do that, while having the printer keep track of the last "type" that it saw, in order to decide if headers should be printed or not.

juanvallejo commented on pkg/kubectl/cmd/resource/get.go:
```diff
@@ -320,145 +345,51 @@ func (options *GetOptions) Run(f cmdutil.Factory, cmd *cobra.Command, args []str
+		ListMeta: metav1.ListMeta{},
+	}
+	for _, info := range infos {
+		list.Items = append(list.Items, info.AsInternal())
```
This was being done in the giant print loop that existed in this file before.
We want the internal version of objects since that is what the printHandlers recognize.

juanvallejo commented on pkg/printers/humanreadable.go:
```diff
@@ -291,16 +293,43 @@ func printHeader(columnNames []string, w io.Writer) error {
+			return utilerrors.NewAggregate(errs)
+		}
+
+		if len(h.options.SortBy) > 0 && len(items) > 1 {
```
This does not work. Not sure why yet, but still looking into it.

juanvallejo commented on pkg/printers/humanreadable.go:
```diff
@@ -291,16 +293,43 @@ func printHeader(columnNames []string, w io.Writer) error {
+	}
+
+	if h.options.WithKind {
+		gk := GetObjectGroupKind(obj, scheme.Scheme)
```
Does not work yet. Currently prints `<unknown>/resource-name`. Any help here is appreciated

juanvallejo commented on pkg/printers/humanreadable.go:
```diff
@@ -320,6 +349,11 @@ func (h *HumanReadablePrinter) PrintObj(obj runtime.Object, output io.Writer) er
 		includeHeaders := h.lastType != t && !h.options.NoHeaders
+
+		if h.lastType != nil && h.lastType != t && !h.options.NoHeaders {
+			fmt.Fprintf(output, "\n")
```
Spaces between resource kind groups

juanvallejo commented on pkg/printers/humanreadable_flags.go:
```diff
@@ -28,13 +28,14 @@ import (
+	ColumnLabels *[]string
+
+	// get.go-specific values
+	NoHeaders bool
```
How should we best differentiate values that are command-specific, and can be bound to a flag, vs values that are not set via a flag?

soltysh replied:
Comment is the only one that comes to my mind.

juanvallejo commented on pkg/printers/printers.go:
```diff
@@ -115,7 +115,7 @@ func GetStandardPrinter(typer runtime.ObjectTyper, encoder runtime.Encoder, deco
 
 	case "wide", "":
-		humanPrintFlags := NewHumanPrintFlags(options.Kind, options.NoHeaders, options.WithNamespace, options.AbsoluteTimestamps)
+		humanPrintFlags := NewHumanPrintFlags(options.NoHeaders, options.WithNamespace, options.AbsoluteTimestamps)
```
This is actually not hit anymore, since get.go asks for the printer directly from the flags. Will remove

deads2k commented on pkg/kubectl/cmd/resource/get.go:
```diff
@@ -46,11 +46,14 @@ import (
 
 // GetOptions contains the input to the get command.
 type GetOptions struct {
+	*printers.HumanPrintFlags
```
no anonymous struct inclusion.  Just say no!

deads2k commented on pkg/kubectl/cmd/resource/get.go:
```diff
@@ -196,12 +214,20 @@ func (options *GetOptions) Complete(f cmdutil.Factory, cmd *cobra.Command, args
 	}
 	options.Sort = len(isSorting) > 0
 
+	options.OutputFormat, err = cmd.Flags().GetString("output")
```
this looks wrong. Why wasn't it bound?

deads2k commented on pkg/kubectl/cmd/resource/get.go:
```diff
@@ -320,145 +345,51 @@ func (options *GetOptions) Run(f cmdutil.Factory, cmd *cobra.Command, args []str
 
-	sorting, err := cmd.Flags().GetString("sort-by")
+	showKind := options.ShowKind || resource.MultipleTypesRequested(args) || cmdutil.MustPrintWithKinds(objs, infos, nil)
+	options.HumanPrintFlags.ShowLabels = &showKind
```
This looks wrong.  Mutation like this seems like it belongs in complete.

deads2k commented on pkg/kubectl/cmd/resource/get.go:
```diff
@@ -320,145 +345,51 @@ func (options *GetOptions) Run(f cmdutil.Factory, cmd *cobra.Command, args []str
+	showKind := options.ShowKind || resource.MultipleTypesRequested(args) || cmdutil.MustPrintWithKinds(objs, infos, nil)
+	options.HumanPrintFlags.ShowLabels = &showKind
+
+	printer, matches, err := options.HumanPrintFlags.ToPrinter(options.OutputFormat)
```
I should be able to get a printer in complete.  If you can't, something has gone wrong.

soltysh commented on pkg/kubectl/cmd/resource/get.go:
```diff
@@ -320,145 +345,51 @@ func (options *GetOptions) Run(f cmdutil.Factory, cmd *cobra.Command, args []str
 	}
 
-	if nonEmptyObjCount == 0 && !options.IgnoreNotFound {
-		fmt.Fprintln(options.ErrOut, "No resources found.")
```
We can't loose this message. I bet some folks will expect it to appear.

soltysh commented on pkg/kubectl/cmd/resource/get.go:
```diff
@@ -320,145 +345,51 @@ func (options *GetOptions) Run(f cmdutil.Factory, cmd *cobra.Command, args []str
-	if nonEmptyObjCount == 0 && !options.IgnoreNotFound {
-		fmt.Fprintln(options.ErrOut, "No resources found.")
+	w := printers.GetNewTabWriter(options.Out)
+	if err := printer.PrintObj(convertedList, w); err != nil {
```
@deads2k what do you think if we change the signature of this method to be `PrintObj(out io.Writer, objs ...runtime.Object)` ? So that we can get rid of this nasty hacks above?

deads2k replied:
> @deads2k what do you think if we change the signature of this method to be PrintObj(out io.Writer, objs ...runtime.Object) ? So that we can get rid of this nasty hacks above?

Did the visitor ever get put back in place?  My initial look at this pull found that this problem appeared because it changed a visitor pattern to an aggregation pattern.

deads2k replied:
> PrintObj(out io.Writer, objs ...runtime.Object) ? So that we can get rid of this nasty hacks above?

And I'm against. That is going to encourage people to think they can pass an entire list through and they'll write printers that don't work and attempt to do things like sorting which have to happen above a printer.

soltysh replied:
> sorting which have to happen above a printer.

Maybe I misunderstood you, but I thought that sorting should be part of the printing, should it not?

juanvallejo replied:
> Did the visitor ever get put back in place? My initial look at this pull found that this problem appeared because it changed a visitor pattern to an aggregation pattern.

We were not visiting here before, we had a slice of runtime.Objects that was passed to the sorter, the slice was sorted in place, and then looped over. For each object, a PrinterForMapping was obtained, and each object was passed to that printer's PrintObj method.

juanvallejo replied:
If we are doing sorting at the command level, we could keep the slice of runtime.Objects that was here before, sort it in place, then loop over each object and pass it to the printer obtained outside of the loop via `PrintFlags#ToPrinter`, and let the printer keep track internally of the last type that it saw, deciding whether or not to print headers.

soltysh commented:
@juanvallejo the most important bit I just talked with David, no sorting in printers. That should simplify a lot for us. But we need to straighten the name so the name change - yes, but don't move it under printers

juanvallejo commented on pkg/kubectl/cmd/resource/get.go:
```diff
@@ -320,145 +345,51 @@ func (options *GetOptions) Run(f cmdutil.Factory, cmd *cobra.Command, args []str
+
+	printer, matches, err := options.HumanPrintFlags.ToPrinter(options.OutputFormat)
+	if !matches {
+		return fmt.Errorf("unable to match a printer to handle current print options")
```
Should we have a fallback here for now? (NewHumanReadablePrinter() ... ) or a different error message?

juanvallejo commented on pkg/kubectl/cmd/resource/get.go:
```diff
@@ -221,6 +235,51 @@ func (options *GetOptions) Complete(f cmdutil.Factory, cmd *cobra.Command, args
+
+	printer, matches, err := options.HumanPrintFlags.ToPrinter(options.OutputFormat)
+	if !matches {
+		return fmt.Errorf("unable to match a printer to handle current print options")
```
@soltysh I am still not sure how to handle this error better / if we should fallback instead?

juanvallejo replied:
For now (at least for this prototype) I fallback to using `cmdutil.PrinterForOptions(cmdutil.ExtractPrintOptions(cmd, ...))`.

Once we wire printers at the command-level, that might no longer be an option.

If no printer were to match, then I'd consider that a programmer error, but not sure how to best present that through an error in this spot.

juanvallejo commented on pkg/kubectl/cmd/resource/get.go:
```diff
@@ -320,40 +371,24 @@ func (options *GetOptions) Run(f cmdutil.Factory, cmd *cobra.Command, args []str
-	printer = nil
-	var lastMapping *meta.RESTMapping
-	w := printers.GetNewTabWriter(options.Out)
+	options.ShowKind = options.ShowKind || resource.MultipleTypesRequested(args) || cmdutil.MustPrintWithKinds(objs, infos, sorter)
```
@deads2k 
> This looks wrong. Mutation like this seems like it belongs in complete.

Because this depends on `cmdutil.MustPrintWithKinds`, and that method depends on sorted infos, I would have to move the resource builder logic, as well as sorting to the Complete method, which I believe would have to be gated in the event of `--watch` (or reconciled with the resource builder logic from the `options.watch` method)

soltysh replied:
No, `MustPrintWithKinds` is logically similar to `MultipleTypesRequested` in that it verifies if the returned types are different. Can you think of an example where the requested type will be a single type but server will return multiple different? I can't off top of my head, so I'd limit this to the two conditions and move it to `Complete`.

juanvallejo commented:
@soltysh @deads2k Thanks for the review, most comments addressed. There are still a couple of questions.

@soltysh 
>  the most important bit I just talked with David, no sorting in printers. That should simplify a lot for us. But we need to straighten the name so the name change - yes, but don't move it under printers

Done. Renamed the `sorting_printer.go` to `sorter.go` (as well as its test).

I switched from making a single list to iterating over the slice of sorted runtime.Objects (like we were doing before), but vastly simplified the loop.

Also, the `ToPrinter` method is now called from the Complete() method.

juanvallejo commented:
Additionally, I have moved the call to `pkg/printers/internalversion#AddHandlers` to `humanreadable_flags.go`.

juanvallejo commented on pkg/printers/printers.go:
```diff
@@ -124,11 +124,25 @@ func GetStandardPrinter(typer runtime.ObjectTyper, encoder runtime.Encoder, deco
-		fallthrough
-	case "":
+	//case "wide", "":
+	//	humanPrintFlags := NewHumanPrintFlags(options.NoHeaders, options.WithNamespace, options.AbsoluteTimestamps)
```
Had to remove this case as depending on `pkg/printers/humanprintflags` restores the import cycle between the `printers` package and `printers/internalversion` (printers -> humanprintflags -> internalversion -> printers).

soltysh commented on pkg/printers/humanprintflags/humanreadable_flags.go:
```diff
@@ -14,12 +14,14 @@ See the License for the specific language governing permissions and
 */
 
-package printers
+package humanprintflags
```
That's not acceptable, we can't move just this file. We need to step back and re-think the printers package.

deads2k replied:
> I've talked with David, create a new pancake pkg/kubectl/printers and put there all command facing items there. Currently that'll be all the flags.go files.

Human readable printers need to stay in get.go for now.  It's a snowflake for get.go.  The rest can move.

soltysh replied:
I wouldn't agree, HRP can move the same way as the rest, especially that I'm talking *only* about the flags for printers. HRP will be used in more commands than just get.go

soltysh commented on pkg/kubectl/cmd/resource/get.go:
```diff
@@ -221,6 +235,51 @@ func (options *GetOptions) Complete(f cmdutil.Factory, cmd *cobra.Command, args
+				humanPrinter.EnsurePrintWithKind(kind)
+			}
+
+			return humanPrinter.PrintObj(info.AsInternal(), out)
```
Yeah, I was wondering about the cast, but let's leave it as is now. I doubt there's a better approach, currently, but then over the curse of time we'll be able to fix it.

juanvallejo commented on pkg/printers/printers.go:
```diff
@@ -33,30 +33,6 @@ func GetStandardPrinter(typer runtime.ObjectTyper, encoder runtime.Encoder, deco
 
-	case "json", "yaml":
-		jsonYamlFlags := NewJSONYamlPrintFlags()
-		p, matched, err := jsonYamlFlags.ToPrinter(format)
```
@soltysh @deads2k moved all printflags to `pkg/printers`.
Adding the dependency on the new `pkg/printers/flags` package here caused an import cycle.

soltysh replied:
See https://github.com/juanvallejo/kubernetes/pull/5 for what I've meant. I've also included a few minor code changes I'd like to see :)

juanvallejo replied:
(`pkg/kubectl/printers` depends on `pkg/printers/internalversion` because we use that pkg to add printHandlers in the `ToPrinter` method of humanreadable_flags.go)

soltysh commented on pkg/kubectl/cmd/resource/get.go:
```diff
@@ -364,86 +393,16 @@ func (options *GetOptions) Run(f cmdutil.Factory, cmd *cobra.Command, args []str
-			printOpts := cmdutil.ExtractCmdPrintOptions(cmd, printWithNamespace)
-			// if cmd does not specify output format and useOpenAPIPrintColumnFlagLabel flag is true,
-			// then get the default output options for this mapping from OpenAPI schema.
-			if !cmdSpecifiesOutputFmt(cmd) && useOpenAPIPrintColumns {
```
I've just realized this should be handled properly.

soltysh commented on pkg/kubectl/cmd/resource/get.go:
```diff
@@ -364,86 +393,16 @@ func (options *GetOptions) Run(f cmdutil.Factory, cmd *cobra.Command, args []str
-			objToPrint = original
-		}
-		if err := printer.PrintObj(objToPrint, w); err != nil {
-			if !errs.Has(err.Error()) {
```
We'll need that error de-duplication logic back as well.
